src/lua/datetime.lua

```diff
@@ -0,0 +1,611 @@
+            nsec = fraction * 1e9
+        elseif fraction ~= 0 then
+            error('only integer values allowed in timestamp '..
+                  'if nsec, usec, or msecs provided', 2)
+        end
+        hms = true
+    end
+
+    local offset = obj.tzoffset
+    if offset ~= nil then
+        offset = get_timezone(offset)
+        check_range(offset, -720, 720, offset)
+    end
+
+    if obj.tz ~= nil then
+        nyi('tz')
+    end
+
+    -- .year, .month, .day
+    if ymd then
+        y = y or 1970
+        M = M or 1
+        d = d or 1
+        if d < 0 then
+            d = builtin.tnt_dt_days_in_month(y, M)
+        end
+        dt = builtin.tnt_dt_from_ymd(y, M, d)
+    end
+
+    -- .hour, .minute, .second
+    local secs = 0
+    if hms then
+        secs = (h or 0) * 3600 + (m or 0) * 60 + (s or 0)
+    end
+
+    return datetime_new_dt(dt, secs, nsec, offset or 0)
+end
+
+--[[
+    Convert to text datetime values
+
+    - datetime will use ISO-8601 format:
+        1970-01-01T00:00Z
+        2021-08-18T16:57:08.981725+03:00
+]]
+local function datetime_tostring(self)
+    check_date(self, 'datetime.tostring()')
+
+    local buff = ffi.new('char[?]', TOSTRING_BUFSIZE)
+    local len = builtin.datetime_to_string(buff, TOSTRING_BUFSIZE, self)
+    assert(len < TOSTRING_BUFSIZE)
+    return ffi.string(buff)
+end
+
+--[[
+    Dispatch function to create datetime from string or table.
+    Creates default timeobject (pointing to Epoch date) if
+    called without arguments.
+]]
+local function datetime_from(o)
+    if o == nil or type(o) == 'table' then
+        return datetime_new(o)
+    end
+end
+
+--[[
+    Create datetime object representing current time using microseconds
+    platform timer and local timezone information.
+]]
+local function local_now()
+    local d = datetime_new_raw(0, 0, 0)
+    builtin.datetime_now(d)
+    return d
+end
+
+--[[
+    Return table in os.date('*t') format, but with timezone
+    and nanoseconds
+]]
+local function datetime_totable(self)
+    local secs = local_secs(self) -- hour:minute should be in local timezone
+    local dt = local_dt(self)
+
+    return {
+        year = builtin.tnt_dt_year(dt),
+        month = builtin.tnt_dt_month(dt),
+        yday = builtin.tnt_dt_doy(dt),
+        day = builtin.tnt_dt_dom(dt),
+        wday = (ffi.cast('int32_t', builtin.tnt_dt_dow(dt)) + 1) % 7,
+        hour = math_floor((secs / 3600) % 24),
+        min = math_floor((secs / 60) % 60),
+        sec = secs % 60,
+        isdst = false,
+        nsec = self.nsec,
+        tzoffset = self.tzoffset,
+    }
+end
+
+local function datetime_update_dt(self, dt, new_offset)
+    local epoch = local_secs(self)
+    local secs_day = epoch % SECS_PER_DAY
+    epoch = (builtin.tnt_dt_rdn(dt) - DAYS_EPOCH_OFFSET) * SECS_PER_DAY +
+            secs_day
+    self.epoch = utc_secs(epoch, new_offset)
+end
+
+local function datetime_ymd_update(self, y, M, d, new_offset)
+    if d < 0 then
```

Comment:
The RFC says it should work only for -1, not for any negative value.

Reply:
Caller checks that there is valid day number or -1, so essentially it works only for -1.
(But yes, I intentionally have left it a little bit wider than requested, because allowing only -1 is making no much sense. But it's well hidden and protected againts to in outer API checks)

Reply:
Please, stick to the RFC. It says it works only for -1.

> only -1 is making no much sense

Well, it makes for me. Moreover, making -2 and others behave like -1 looks like a bug. I would rather expect -2 be the next to last day, -3 the next to the next to the last day and so on.

Reply:
-2 or -3 will not pass range check here:
```lua
    local d = obj.day
    if d ~= nil then
        check_range(d, 1, 31, 'day', -1)
        ymd = true
    end
```